### ISSUE TITLE
Could you please add a simple program to write a stream of random values to stdout

### ISSUE THREAD

**Issue by jirka-h** [hahnjo/ranluxpp#1]
Hi Jonas,

thanks a lot for starting this project! Could you please add info on how to compile the project to README.md? Simple 

```
cmake CMakeLists.txt
```

will do. 

Also, could you please add a simple program to write a stream of random values to stdout (with fwrite) so that the output could be checked for example with PractRand?

```
ranluxpp-test | practrand-RNG_test stdin -tlmax 64G -multithreaded
```

On Fedora, CentOS and RHEL, you can install PractRand with
```
dnf install practrand
```

Thanks a lot!
Jirka

**Comment by hahnjo** [comment]
Thanks for letting me know that there's a `practrand` package in the RHEL world! I looked at PractRand in the past, but eventually gave up due to its (basically non-existent) build system.

However, my understanding of `practrand-RNG_test stdin` is that it expects random bit data. After talking to theory people, this is not how RANLUX / RANLUX++ should be used - you obtain the strong properties when you use 48 bits, for example to produce one floating point number (or 24 bits as in the original RANLUX). For that reason, I'm unlikely to add a program that makes it easy to use the generator incorrectly.

What I would be interested in is using PractRand's tests via the library interface. But it seems that the `practrand` package only contains the binaries, not the libraries and header files?

**Comment by jirka-h** [comment]
Hi Jonas,

I was missing Practrand binary package so much that I have decided to create one:-) 

I was considering preparing a devel package as well, but I have dropped the idea for several reasons:

- There is absolutely no upstream support. The software was never developed as a library
- Current interface is difficult to use (See http://pracrand.sourceforge.net/Tests_usage.txt and [1])
- On current HW with many CPUs available, passing the data through the pipe is almost as fast as adding your RNG directly to PractRand. The pipe version is utilizing two CPUs (one for RNG, the other for the test) which cancels out the extra cost of writing and reading from the pipe. Try to compare runtime of `practrand-RNG_output jsf64 inf | practrand-RNG_test stdin64 -tlmax 4G` against `practrand-RNG_test jsf64 -tlmax 4G`. On my laptop, the runtime is almost the same. 
- Others have reached the same conclusion - see for example https://www.pcg-random.org/posts/how-to-test-with-practrand.html

> You obtain the strong properties when you use 48 bits, for example to produce one floating point number (or 24 bits as in the original RANLUX).

My understanding is that 48-bits used to produce one double value has to pass the randomness test. So if you take 48 bits and write them as a binary stream, it has to pass the empirical tests. This is how RNGs producing double values are tested. Am I missing something here? If you disagree, could you please elaborate on why you think this is not the correct approach? 

Liebe Grüße :-)
Jirka

[1] However, it tends to require more work as you have to understand some of the subtleties of the internals of PractRand.  The easiest way may be to edit the "dummy" RNG in RNG_test.cpp, but other ways are possible too.

**Comment by hahnjo** [comment]
> There is absolutely no upstream support. The software was never developed as a library

True, that's what I also experienced :cry: 

> > You obtain the strong properties when you use 48 bits, for example to produce one floating point number (or 24 bits as in the original RANLUX).
> 
> My understanding is that 48-bits used to produce one double value has to pass the randomness test. So if you take 48 bits and write them as a binary stream, it has to pass the empirical tests.

*If* the reading end takes blocks of 48 bits, and *if* it treats the result as an equally distributed number from 0 to 2 ** 48 - 1, then I tend to agree. However, PractRand's `stdin` seems to read 8 bits at a time and while there are other sizes available, 48 bits is not one of them.

The reason why blocks of 48 bits are so important for RANLUX is that the bits in the output numbers are not equally distributed on their own. The most intuitive explanation is that the internal LCG has a state from 0 to m = 2 ** 576 - 2 ** 240 + 1, while the RANLUX numbers plus the carry bit have a state size of (2 ** 576) * 2. This means not all of the latter states are possible to reach, and practically speaking (in my understanding) leaving out some states implies that the bits are not equally likely to be 0 and 1.

> This is how RNGs producing double values are tested. Am I missing something here? If you disagree, could you please elaborate on why you think this is not the correct approach?

No, for example TestU01 takes `double` values as input. Then you don't care about the distribution of individual bits, but you need the whole number to be equally distributed in [0, 1). And that we get from RANLUX where the 24 bit numbers are distributed in [0, 2 ** 24) and, by extensions, 48 bit numbers in [0, 2 ** 48). Remember, it's not a cryptographic RNG where all bits are random.

**Comment by jirka-h** [comment]
> This means not all of the latter states are possible to reach, and practically speaking (in my understanding) leaving out some states implies that the bits are not equally likely to be 0 and 1.

Thanks for the explanation. It makes sense. It also means that RANLUX double output is not completely evenly distributed on the interval [0, 1). But it's not a cryptographic RNG so it's fine. 

I have reviewed TestU01 and it requires each generator to provide two methods:
- function GetU01 that returns values in the interval [0, 1) 
- and a function GetBits that returns a block of 32 bits.

For RANLUX implementation in TestU01, there is `Ranlux_Bits`  which returns the first most significant 32 bits:
```
#define unif01_NORM32  4294967296.0
static unsigned long Ranlux_Bits (void *vpar, void *vsta)
{
   return (unsigned long) (unif01_NORM32 * Ranlux_U01 (vpar, vsta));
}
```
I have profiled Small Crush battery with the gprof
```
  gen = ucarry_CreateRanlux (576, 0);
  bbattery_SmallCrush (gen);
```
and Ranlux_Bits is indeed heavily used. So at least some of TestU01's tests are not working with double values but with the bits representation. 

I think the same approach can be used for RANLUX++ - take the first 32-bits from 48 bits and pass them to PractRand using stdin32 test (`practrand-RNG_test stdin32`).

I could also argue that using whole 48-bits and passing them to PractRand stdin test (which can treat them as separate 8-bit values) is in fact more strict test. It could be actually interesting to run it - we know that individual bits are not completely random, so it will be intriguing to check if PractRand can detect these theoretical flaws. I will try to set it up.

**Comment by hahnjo** [comment]
> > This means not all of the latter states are possible to reach, and practically speaking (in my understanding) leaving out some states implies that the bits are not equally likely to be 0 and 1.
> 
> Thanks for the explanation. It makes sense. It also means that RANLUX double output is not completely evenly distributed on the interval [0, 1).

No, that doesn't follow directly because there are 12 numbers generated from each LCG state. And while not all bit combinations are possible for every single number, that doesn't imply the stream of numbers cannot be evenly distributed.

> I have reviewed TestU01 and it requires each generator to provide two methods:
> 
>     * function GetU01 that returns values in the interval [0, 1)
> 
>     * and a function GetBits that returns a block of 32 bits.

Hm, interesting: I've been using `unif01_CreateExternGen01` and that also seems to add a `GetBits` function...

> For RANLUX implementation in TestU01, there is `Ranlux_Bits` which returns the first most significant 32 bits:
> 
> ```
> #define unif01_NORM32  4294967296.0
> static unsigned long Ranlux_Bits (void *vpar, void *vsta)
> {
>    return (unsigned long) (unif01_NORM32 * Ranlux_U01 (vpar, vsta));
> }
> ```

(but at least it's extracting the full U01 (24-bit) number from RANLUX first, then scaling by 2 ** 32)

> 
> I have profiled Small Crush battery with the gprof
> 
> ```
>   gen = ucarry_CreateRanlux (576, 0);
>   bbattery_SmallCrush (gen);
> ```

(note that to match the luxury level of RANLUX++, you'll need `p = 2048`)

> and Ranlux_Bits is indeed heavily used. So at least some of TestU01's tests are not working with double values but with the bits representation.
> 
> I think the same approach can be used for RANLUX++ - take the first 32-bits from 48 bits and pass them to PractRand using stdin32 test (`practrand-RNG_test stdin32`).

To match the above snippet, you should extract the `double` and then scale by 2 ** 32. Which is probably fine for all empirical tests that you can run in your lifetime, since you have 16 bits of "room" to cover up any problem in the original bits.

> I could also argue that using whole 48-bits and passing them to PractRand stdin test (which can treat them as separate 8-bit values) is in fact more strict test. It could be actually interesting to run it - we know that individual bits are not completely random, so it will be intriguing to check if PractRand can detect these theoretical flaws. I will try to set it up.

To be honest, I wouldn't expect any reasonably sized test to find problems...

**Comment by jirka-h** [comment]
> No, that doesn't follow directly because there are 12 numbers generated from each LCG state. And while not all bit combinations are possible for every single number, that doesn't imply the stream of numbers cannot be evenly distributed.

I agree. But you will (at least in theory) discover the flaw with [Spectral test](https://en.wikipedia.org/wiki/Spectral_test). If you start placing 12 numbers into a 12-dimensional space, you will see some voids - space with no points. Good RNG is expected to pass such a test. But again, for RANLUX is deviation is very small and it's passing empirical tests. 

> To be honest, I wouldn't expect any reasonably sized test to find problems...
I agree. So far, I have only tested RNG by Sibidanov (without conversion LCG -> RANLUX state) and it has passed both stdin and stdin64 tests up to 64 TB (I haven't run tests any longer). Command-line was:
```
practrand-RNG_test stdin -tlmax 64T -multithreaded
practrand-RNG_test stdin64 -tlmax 64T -multithreaded
```
From 576-bit LCG state, I have also tested 
- most significant byte out of a 576-bit word
- least significant byte out of a 576-bit word

and I was hoping to discover some irregularities, but it has passed Practrand stdin8 tests without any problems (tested up to 2TB). Advancing LCG state by 2048 steps seems to be a very efficient way how to mask any irregularities.

I haven't tested so far runtime penalty for converting LCG to RANLUX state - I plan to do it when I'm back from vacation. I assume it will be small and then it's the right thing to do. We know it will improve the quality of RNG - but it seems to be very challenging to spot the difference by any empirical tests on today's commodity HW.
